decentraland/unity-explorer
# Feat: Worlds Environment

fix #532 

Adds Terrain generation for Worlds (`.dcl.eth`) 
- created generator for worlds that adapts to its size and adds padding with several parcels
- added boundary colliders spawning
- managed switch terrains when chaning realms
- extracted terrain creation and configuration logic into TerrainFactory
- extracted generations related to TerrainChunk into separate class
- changed measures to be scoped 

### NOTE:
This PR has small bugs, some performance and optimization issues that we need to extract into separate issue and handle lately.. This includes (🚀 - **Performance**, 🚧 - **Refactoring**,  🐞- **Bug** ):
- 🚀 : old `DigHolesAsync `method is not used (its usage is commented). We leave it for now as a reference. We need to revisit it and apply Jobs to the presented hole generation logic. 
- 🚀 : `WorldTerrainGenerator` need proper resource management for re-using assets (pools, caches, etc.)
- **Refactoring**:  `WorldTerrainGenerator` and `TerrainGenerator` probably can be merged in one class with different strategies for respective logic.
- 🚧 :  `WorldTerrainGenerator` has almost copy-pasted methods from `TerrainGenerator`, such as `SetupEmptyParcelDataAsync`...
- 🚧 :  `TerrainChunkDataGenerator` should return arrays and not manage logic of `.Set` on the `TerrainData`. Decorators can be applied for Caching and Logging.
- 🚧 :  `TerrainChunkDataGenerator`, would be nice to extract each generation algorithm (tree, holes) into a strategy and have single-treaded version of it for easier debugging (if possible). 
- 🐞 : Can't generate terrain if padding == 0
- 🐞 : Don't generate trees on padded parcels (for GCity at least)
- 🐞 : Small vertical gap between GCity terrain and padded parcels through which you can see the sea

### Basic logic
Keep in mind, that at the end Terrain is composed by a set of TerrainObjects (Unity specific GameObjects with Terrain component), that we refer as Chunks. Generation algorithm is like that:
- calculate TerrainModel for current World (based on World parcels). 
- spawn non-terrain GameObjects (Cliffs, Colliders, Wind, Water, Root)
- calculate TerrainData for each TerrainChunk (related heavy logic is inside `TerrainChunkDataGenerator.cs`)
- spawn TerrainObject for each TerrainChunk (based on TerrainData)

# How to test
- Go in Genesis City and verify terrain is same as on main (trees, cliffs, wind)
- type `/goto olvra.dcl.eth` or any other world
- verify that new terrain is generated
- jump back to genesis city and verify that its terrain is back
- jump between different worlds and check there is no errors

- additionally - test with 2 players jumping in the same World and verify trees and rocks location same for both of you
